### PR TITLE
High Contrast hint on settings portal

### PR DESCRIPTION
**Pull request by dominichayesferen**
# Introduction
High Contrast is a feature where, much as the name implies, the contrast of applications that support the feature is increased, whether it be adding more opacity to lines in applications, or in the rare case changing the entire palette of an application to be of the maximum possible colour contrast. Irregardless of implementation, this hint has existed for quite a long time in the UNIX space, as well as existing in most other Operating Systems, and this pull request gives this long-standing accessibility hint a standardised location for both existing and future uses of high contrast to come.

# Details
A new key on the settings portal, `high-contrast`, would be introduced into the `org.freedesktop.appearance` namescape. This key will use a boolean value.

The design of this key is based on every existing instance of a high contrast switch - in all platforms I know of, including those in the UNIX space who provide this value in a non-standardised manner, it has always been a simple boolean value which, if True, enables high contrast, but otherwise restores normal application styling if False.

This serves as a hint, just like its non-standardised variations, informing applications of the desire for contrast to be increased - applications are absolutely free to ignore it if, for example, they believe their default application appearance is of high enough contrast.

The hope for a standardised high contrast switch is that it will allow toolkits' existing high contrast options to be easily accessible from other Desktop Environments, such as GTK's high contrast option outside of GNOME, as well as encouraging programs, natively supported on XDG-compliant platforms, that technically have high contrast support programmed in, such as Mozilla Firefox and the Chromium Project, to expose this functionality on XDG-compliant platforms too.

# Acks

- [x] @snwh (GNOME)
- [x] @sonnyp (GNOME)
- [x] @JoshStrobl (Budgie)
- [x] @CarlSchwan (KDE)
- [x] @mmstick (Cosmic)
- [x] @danirabbit (Elementary)
- [x] @emilio (Gecko (Firefox's) Browser Engine)
- [x] @nothingneko (Helium)

# Implementations of High Contrast

- [ ] KDE Plasma: https://invent.kde.org/plasma/breeze/-/merge_requests/397
- GNOME, and all GTK-powered Desktop Environments: High Contrast boolean-switch in Settings
- Firefox: Currently only available on GNOME

Technically also exists in Chromium, although is currently inaccessible on Linux.

# Implementations using this standard

None yet.

### CODE REVIEW

**Comment by GeorgesStavracas** [comment]
I think this is pretty uncontroversial as is, thanks for proposing it. Let's just check if anyone objects to this proposal.

**Comment by GeorgesStavracas** [comment]
Just taking the list of ACKs from #815:

- [ ] @alice-mkh (GNOME)
- [ ] @JoshStrobl (Budgie)
- [ ] @Sodivad (KDE)
- [ ] @jackpot51 (Cosmic)
- [ ] @danirabbit (Elementary)

**Comment by JoshStrobl** [comment]
Ack on the Budgie side. Seems like a pretty obvious key to me.

**Comment by dominichayesferen** [comment]
I'll also ping @emilio in here in advance, as I remember them participating in the accent colour portal MR discussion, as a pseudo-rep of sorts for Gecko and thus presumably Firefox - would you be able to, by chance, be a, or alternatively get a, representative from Mozilla Firefox's maintainers in here by chance to also see if they suggest anything to add as part of a standardised high contrast hint?

As for Chromium, not sure about how to get in touch with getting a representative for their side, for the record.

**Comment by lleyton** [comment]
I wonder if the namespace for the key should be something like `org.freedesktop.accessibility` instead of `org.freedesktop.appearance`. That's a minor detail though, everything else seems fine to me :p

**Comment by dominichayesferen** [comment]
Follow up from lleyton's above comment:
If the general consensus is in favour of Lleyton's idea, should it be called "org.freedesktop.accessibility" or "org.freedesktop.a11y"?

**Comment by JoshStrobl** [comment]
@dominichayesferen Personally prefer `accessibility` over `a11y` but not fussed either way.

**Comment by GeorgesStavracas** [comment]
I have a philosophical appreciation of Cassidy's ["accessibility features are just features"](https://blog.elementary.io/accessibility-features-are-just-features/) assertion, so, in my humble opinion, `org.freedesktop.appearance` is fine. But it's not a strong opinion.

**Comment by GeorgesStavracas** [comment]
I'm going to leave it up to you all to agree on the naming scheme, I'm fine with all the options presented so far, and my preference for `org.freedesktop.appearance` over others is very faint.

**Comment by danirabbit** [comment]
I don't have super strong opinions here. We haven't historically supported a separate high contrast mode because our goal is to meet [WCAG AA by default](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) and my understanding is that folks with significant vision loss past this level would likely need other types of assistance like magnification or a screen reader.

WCAG also talks about some folks needing low contrast, so I'm not sure if it makes sense to go ahead and add a three-state setting for contrast if there's going to be a contrast setting? Or if it would be better assumed that a low contrast setting would be implemented by a compositor shader or something

**Comment by orowith2os** [comment]
Ack on `appearance`.

> Similarly, as not all Desktop Environments currently employ any form of High Contrast switches, this hint should also be considered optional for the Portal implementers - if you do not want a high contrast option in your Desktop Environment, simply hardcode returning False into your Portal implementation (remember, however, that by doing so third-party applications supporting this feature will never expose their high contrast implementations if this preference is merged and they start following it).

Did any app ever expose it in their UI? GNOME's and KDE's sure didn't. It's probably in the same realm as dark mode; it should be controlled in the system menus, not per-app. 

But that would be getting into design discussions unrelated to this change. Nothing to block this on. I'd like to merge this as-is.

**Comment by dominichayesferen** [comment]
> Ack on `appearance` (as a third party).
> 
> > Similarly, as not all Desktop Environments currently employ any form of High Contrast switches, this hint should also be considered optional for the Portal implementers - if you do not want a high contrast option in your Desktop Environment, simply hardcode returning False into your Portal implementation (remember, however, that by doing so third-party applications supporting this feature will never expose their high contrast implementations if this preference is merged and they start following it).
> 
> Did any app ever expose it in their UI? GNOME's and KDE's sure didn't. It's probably in the same realm as dark mode; it should be controlled in the system menus, not per-app.
> 
> But that would be getting into design discussions unrelated to this change. Nothing to block this on. I'd like to merge this as-is.

I'm referring to the following:

- Applications: They can choose to read whatever the preference is, and if they want to and the preference returns True they can switch into a state of higher contrast
- Desktops: GNOME, Cinnamon, MATE, Cinnamon, maybe Budgie, and possibly XFCE (not sure though) all expose high contrast, in a non-standardised form, in their System Settings applications (currently under Accessibility).

**Comment by dominichayesferen** [comment]
> I don't have super strong opinions here. We haven't historically supported a separate high contrast mode because our goal is to meet [WCAG AA by default](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) and my understanding is that folks with significant vision loss past this level would likely need other types of assistance like magnification or a screen reader.
> 
> WCAG also talks about some folks needing low contrast, so I'm not sure if it makes sense to go ahead and add a three-state setting for contrast if there's going to be a contrast setting? Or if it would be better assumed that a low contrast setting would be implemented by a compositor shader or something

I'm more than willing to convert this to a tri-state like theme-preference is, but while a part of me says we should just so that we don't need to make a V2 to add it in the future, another part of me says that might be too niche, what with literally no platform I know of having a 'low contrast' option except ancient GTK3 and maybe GTK2.

**Comment by dominichayesferen** [comment]
Actually, given that you could argue theme-preference's "Prefer Light" is of similar niche... probably best for it to have it just on the off chance - although I REALLY don't expect anyone to ever implement it anywhere, though.

I suppose if nobody NACKs that idea, I'll implement it tomorrow or the next day.

**Comment by GeorgesStavracas** [comment]
Then I suggest turning this into a `contrast` string property or somesuch, with 2 documented values: `"default"` and `"high-contrast"`. This will allow documenting a third `"low-contrast"` mode in the future without breaking API.

I don't think it's useful to add low-contrast mode right now. It seems like an unnecessary burden right now, given that apparently nobody implements it.

**Comment by dominichayesferen** [comment]
I'd probably make it the following:
0 - normal
1 - high
2 - low (undocumented until needed?)

**Comment by GeorgesStavracas** [comment]
That would work as well. Just don't document low-contrast right now, we should evaluate adding that later.

**Comment by dominichayesferen** [comment]
*dualstate
ah whatever doesn't matter 'cos I'll probably need to manually squeeze the commits anyway later.

**Comment by dominichayesferen** [comment]
Anyway, how's that for a structure to accommodate lower contrast?

**Comment by dominichayesferen** [comment]
Also just in case there's another person on the off chance that doesn't realise High Contrast exists in GNOME:
image

**Comment by orowith2os** [comment]
> Anyway, how's that for a structure to accommodate lower contrast?

Can't be any better, I like it. 

>  Think of it as similar logic to the reasoning for having a potential colour blindness hint in Portal in the future - letting applications, especially games come into mind, account themselves for any filters in use.

This, and many other settings hidden behind the `org.gnome` namespace, would also be great candidates for including in the Settings portal next, since color blindness might take a bit to design and work on on the toolkit side of things.

**Comment by lleyton** [comment]
> I have a philosophical appreciation of Cassidy's ["accessibility features are just features"](https://blog.elementary.io/accessibility-features-are-just-features/) assertion, so, in my humble opinion, `org.freedesktop.appearance` is fine. But it's not a strong opinion.

I wasn't aware of that article, but I'll admit the argument is quite compelling. `org.freedesktop.appearance` makes sense then.

> Anyway, how's that for a structure to accommodate lower contrast?

Seems good with me :)

> This, and many other settings hidden behind the org.gnome namespace, would also be great candidates for including in the Settings portal next, since color blindness might take a bit to design and work on on the toolkit side of things.

Agreed.

**Comment by dominichayesferen** [comment]
I feel that y'all will be able to handle discussion about this merge request in particular while I'm absent from it in lieu of IRL stuff and focusing on college, so to allow for that to potentially happen, it's time I undraft this.

**Comment by emilio** [comment]
I had missed the ping above, but this seems fine to me as well. Thanks for working on this!

**Comment by dominichayesferen** [comment]
> I had missed the ping above, but this seems fine to me as well. Thanks for working on this!

Should I count that as an ack on Mozilla's side, or...?

**Comment by emilio** [comment]
Sure

**Comment by snwh** [comment]
I would concur with @GeorgesStavracas that overall this seems pretty uncontroversial. I'm reminded of the [prefers-contrast](https://www.w3.org/TR/mediaqueries-5/#prefers-contrast) media query in CSS, which has already been mentioned, and would also agree on doing a 3 state key for `contrast` that somewhat mirrors the WCAG spec:

- `default`: no preference, or a deference to the current state
- `high`: hint for an active option that increases contrast 
- `low`: hint for an active option that reduces contrast

I do think even if not all the stakeholders fully implement something like low contrast, it'd be good to have it in and documented just so it can be forward-looking, especially if this is an accommodation that could be made in future for a subset of users.

**Comment by dominichayesferen** [comment]
@sonnyp @snwh Can I confirm that you two are ACKing this MR, given your messages and who you may be representing?

**Comment by dominichayesferen** [comment]
Alright, with that, there's only a few ACKs left I would think.

**Comment by CarlSchwan** [comment]
The proposal looks sensible to me.

ack from KDE

**Comment by snwh** [comment]
> @sonnyp @snwh Can I confirm that you two are ACKing this MR, given your messages and who you may be representing?

Yeah, weighing in from GNOME. :+1:

**Comment by mmstick** [comment]
ack from COSMIC

**Comment by danirabbit** [comment]
Ack 😊 thanks everyone

**Comment by dominichayesferen** [comment]
Wow, this actually reached 'full house' in terms of its ACKs, thanks everyone from basically all of the corners of FreeDesktop and beyond for providing your concerns, and eventually ACKs! 💙 

I suppose this means this is ready for merging, then, but that's on those who can merge. In the meantime, if there are any MRs made in toolkits or desktop environments, with this proposal in mind, please drop them in this discussion so I can add them to the main message.

**Review by GeorgesStavracas — APPROVED** [review]
Please fix the commit history to have a single, properly formatted commit (see the main branch for commit message guidelines).

**Comment by nothingneko** [comment]
Ack from Helium. Let's do this

**Comment by dominichayesferen** [comment]
Alright, gotta remember how to flatten commits properly, so that I don't mess up and accidentally auto-close this PR.

**Comment by nothingneko** [comment]
I think you can do it from the GitHub UI. Should be a button for a maintainer to squash and merge.

**Comment by dominichayesferen** [comment]
@GeorgesStavracas Ready!

**Comment by dominichayesferen** [comment]
Guess that means I'm now officially a FreeDesktop standard author, nice.

**Comment by dominichayesferen** [comment]
oh yeah, @emilio, now it's merged you might be interested in the bug report made in advance: https://bugzilla.mozilla.org/show_bug.cgi?id=1861775

**Comment by hfiguiere** [comment]
Shall we implement it in the -gtk portal as well?

**Pull request merged**
